File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.java/core/src/main/java/de/unistuttgart/hamster/command/impl/CompositeCommandBase.java
package de.unistuttgart.hamster.command.impl;

import de.unistuttgart.hamster.command.Command;
import de.unistuttgart.hamster.command.CompositeCommand;
import de.unistuttgart.hamster.command.Entity;

public abstract class CompositeCommandBase extends CompositeCommand {

    @Override
    public void undo() {
        var subCommands = getSubCommands();
        for (int i = subCommands.size() - 1; i >= 0; i--) {
            Command subCommand = subCommands.get(i);
            subCommand.undo();
        }
    }

    @Override
    public void redo() {
        var subCommands = getSubCommands();
        for (Command subCommand : subCommands) {
            subCommand.redo();
        }
    }

    protected void executeSetProperty(Entity entity, String propertyName, Object oldValue, Object newValue) {
        var command = new SetPropertyCommandImpl();
        command.setEntity(entity);
        command.setOldValue(oldValue);
        command.setNewValue(newValue);
        command.setPropertyName(propertyName);

        getSubCommands().add(command);
        command.execute();
    }

    protected void executeAddReference(Entity entity, String propertyName, Entity entityToAdd) {
        var command = new AddEntityCommandImpl();
        command.setEntity(entity);
        command.setEntityToAdd(entityToAdd);
        command.setPropertyName(propertyName);

        getSubCommands().add(command);
        command.execute();
    }

    protected void executeRemoveReference(Entity entity, String propertyName, Entity entityToRemove) {
        var command = new RemoveEntityCommandImpl();
        command.setEntity(entity);
        command.setEntityToRemove(entityToRemove);
        command.setPropertyName(propertyName);

        getSubCommands().add(command);
        command.execute();
    }

}

File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/src/impl/CompositeCommandBase.cpp
#include "CompositeCommandBase.h"

#include "AddEntityCommandImpl.h"
#include "RemoveEntityCommandImpl.h"

namespace commands {
  void CompositeCommandBase::undo() {
    auto subCommands = getSubCommands();
    for (auto iter = subCommands.rbegin(); iter != subCommands.rend(); ++iter) {
      auto& subCommand = *iter;
      subCommand->undo();
    }
  }

  void CompositeCommandBase::redo() {
    auto subCommands = getSubCommands();
    for (auto& subCommand : subCommands) {
      subCommand->redo();
    }
  }

  void CompositeCommandBase::executeSetProperty(std::shared_ptr<Entity> entity, std::string propertyName, Any oldValue, Any newValue) {
    internalExecuteSetProperty(entity, propertyName, oldValue, newValue);
  }

  SetPropertyCommandImpl& CompositeCommandBase::internalExecuteSetProperty(std::shared_ptr<Entity> entity, std::string propertyName, Any oldValue, Any newValue) {
    auto command = std::make_shared<SetPropertyCommandImpl>();
    command->setEntity(entity);
    command->setOldValue(oldValue);
    command->setNewValue(newValue);
    command->setPropertyName(propertyName);

    addToSubCommands(command);
    command->execute();
    return *command;
  }

  void CompositeCommandBase::executeAddReference(std::shared_ptr<Entity> entity, std::string propertyName, std::shared_ptr<Entity> entityToAdd) {
    auto command = std::make_shared<AddEntityCommandImpl>();
    command->setEntity(entity);
    command->setEntityToAdd(entityToAdd);
    command->setPropertyName(propertyName);

    addToSubCommands(command);
    command->execute();
  }

  void CompositeCommandBase::executeRemoveReference(std::shared_ptr<Entity> entity, std::string propertyName, std::shared_ptr<Entity> entityToRemove) {
    auto command = std::make_shared<RemoveEntityCommandImpl>();
    command->setEntity(entity);
    command->setEntityToRemove(entityToRemove);
    command->setPropertyName(propertyName);

    addToSubCommands(command);
    command->execute();
  }

}
